src/program/error/mod.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 Rigetti Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod leftover;
+mod result;
+mod syntax;
+
+use std::error::Error;
+use std::fmt;
+use std::fmt::Formatter;
+
+use crate::instruction::Instruction;
+use crate::parser::{LexError, ParseError};
+pub use leftover::LeftoverError;
+pub use result::{convert_leftover, disallow_leftover, map_parsed, recover};
+pub use syntax::SyntaxError;
+
+/// Errors that may occur while parsing a [`Program`](crate::program::Program).
+#[derive(Debug, PartialEq)]
+pub enum ProgramError<T> {
+    InvalidCalibration {
+        instruction: Instruction,
+        message: String,
+    },
+    RecursiveCalibration(Instruction),
+    Syntax(SyntaxError),
+    Leftover(LeftoverError<T>),
```

Comment:
Leftover is also a syntax / parsing-related error. Shouldn't that be down at that level?

Reply:
It could be, yeah. I don't have a preference, so I'll make the change.